crates/typst/src/eval/plugin.rs

```diff
@@ -135,7 +138,14 @@ impl Plugin {
     }
 
     fn call_inner(&self, function_name: &str, args: &[&[u8]]) -> StrResult<Vec<u8>> {
-        self.store().data_mut().arg_buffer = args.concat();
+        self.store().data_mut().arg_buffer.clear();
+        for arg in args {
+            self.store()
+                .data_mut()
+                .arg_buffer
+                .write_all(arg)
+                .map_err(|_| EcoString::from("Coudn't write in the plugin arg buffer"))?;
+        }
```

Comment:
This stills concatenate the args into `arg_buffer` though: the number of copies stays the same...

I think what @laurmaedje meant was, don't copy the args into `arg_buffer`. Rather, find a way to put the in the wasm memory directly.

I experimented a bit, and the following design seems to work fine (without changing the plugin protocol):
```diff
struct PersistentData {
    result_data: Vec<u8>,
-    arg_buffer: Vec<u8>,
+    arguments: Vec<Bytes>,
}

// ...

.func_wrap(
    "typst_env",
    "wasm_minimal_protocol_write_args_to_buffer",
    move |mut caller: Caller<PersistentData>, ptr: u32| {
        let memory =
            caller.get_export("memory").unwrap().into_memory().unwrap();
-        let buffer = std::mem::take(&mut caller.data_mut().arg_buffer);
-        memory.write(&mut caller, ptr as _, &buffer).unwrap();
-        caller.data_mut().arg_buffer = buffer;
+        let arguments = std::mem::take(&mut caller.data_mut().arguments);
+        let mut offset = ptr as usize;
+        for arg in arguments {
+            memory.write(&mut caller, offset, arg.as_slice()).unwrap();
+            offset += arg.as_slice().len();
+       }
    },
)

// ...

- fn call_inner(&self, function_name: &str, args: &[&[u8]]) -> StrResult<Vec<u8>> {
-    self.store().data_mut().arg_buffer.clear();
-    for arg in args {
-        self.store()
-            .data_mut()
-            .arg_buffer
-            .write_all(arg)
-            .map_err(|_| EcoString::from("Coudn't write in the plugin arg buffer"))?;
-    }
+ fn call_inner(&self, function_name: &str, args: Vec<Bytes>) -> StrResult<Vec<u8>> {
+        self.store().data_mut().arguments = args;
```

Reply:
Ah yes, that seems smarter indeed!

Reply:
Should be okay now!